packages/agreement-process/src/services/agreementStateProcessor.ts

```diff
@@ -144,3 +169,179 @@ export const suspendedByProducerFlag = (
+    nextState,
+    agreement.data.suspendedByProducer,
+    agreement.data.suspendedByConsumer,
+    newSuspendedByPlatform
+  );
+
+  const updatedAgreement: Agreement = {
+    ...agreement.data,
+    state: finalState,
+    suspendedByPlatform: newSuspendedByPlatform,
+  };
+
+  if (
+    allowedStateTransitions(agreement.data.state).includes(finalState) &&
+    newSuspendedByPlatform !== agreement.data.suspendedByPlatform
+  ) {
+    return match(finalState)
+      .with(agreementState.suspended, () =>
+        toCreateEventAgreementSuspendedByPlatform(
+          updatedAgreement,
+          agreement.metadata.version,
+          correlationId
+        )
+      )
+      .with(agreementState.active, () =>
+        toCreateEventAgreementUnsuspendedByPlatform(
+          updatedAgreement,
+          agreement.metadata.version,
+          correlationId
+        )
+      )
+      .with(agreementState.missingCertifiedAttributes, () =>
+        toCreateEventAgreementPutInMissingCertifiedAttributesByPlatform(
+          updatedAgreement,
+          agreement.metadata.version,
+          correlationId
+        )
+      )
+      .with(agreementState.draft, () =>
+        toCreateEventAgreementPutInDraftByPlatform(
+          updatedAgreement,
+          agreement.metadata.version,
+          correlationId
+        )
+      )
+      .otherwise(
+        () =>
+          void logger.error(
+            `Agreement state transition not allowed from ${agreement.data.state} to ${finalState}`
+          )
+      );
+  } else {
+    logger.error(
+      `Agreement state transition not allowed from ${agreement.data.state} to ${finalState}`
+    );
+  }
+}
+
+function eserviceContainsAttribute(
+  attributeId: AttributeId,
+  eservice: EService
+): boolean {
+  const certified = eservice.descriptors.flatMap(
+    (descriptor) => descriptor.attributes.certified
+  );
+  const declared = eservice.descriptors.flatMap(
+    (descriptor) => descriptor.attributes.declared
+  );
+  const verified = eservice.descriptors.flatMap(
+    (descriptor) => descriptor.attributes.verified
+  );
+
+  const allIds = [...certified, ...declared, ...verified].flatMap((attr) =>
+    attr.map((a) => a.id)
+  );
+
+  return allIds.includes(attributeId);
+}
+
+// eslint-disable-next-line max-params
+export async function computeAgreementsStateByAttribute(
+  attributeId: AttributeId,
+  consumer: CompactTenant,
+  agreementQuery: AgreementQuery,
+  eserviceQuery: EserviceQuery,
+  correlationId: string,
+  logger: Logger
+): Promise<Array<CreateEvent<AgreementEvent>>> {
+  const agreements = await agreementQuery.getAllAgreements({
+    consumerId: consumer.id,
+    agreementStates: updatableStates,
+  });
+
+  const eservices: EService[] = (
+    await Promise.all(
+      agreements
+        .map((agreement) => agreement.data.eserviceId)
+        .map(eserviceQuery.getEServiceById)
+    )
+  ).filter(
+    (eservice): eservice is NonNullable<EService> => eservice !== undefined
+  );
```

Comment:
```suggestion
  const uniqueEServiceIds = [
    ...new Set(agreements.map((a) => a.data.eserviceId)),
  ];
  const eservices: EService[] = await Promise.all(
    uniqueEServiceIds.map(async (id) => {
      const eservice = await eserviceQuery.getEServiceById(id);
      if (!eservice) {
        throw eServiceNotFound(id);
      }
      return eservice;
    })
  );
```
I suggest these two changes to align it with the Scala code:
1. use `uniqueEServiceIds` in order to not retrieve more than once the same eservice;
2. replace the filter with a check in order to throw the error if we don't find the eservice.

P.S.: the whole code inside the map can be replaced with `retrieveEservice` (Which does both the get and the check) when the refactor will be merged in order to simplify the it.

Reply:
Fixed 👍  I left a TODO to replace this with `retrieveEservice`, if this is merged before you can do it in the refactor PR, otherwise I'll merge and do it here